uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Service/LeaveBalanceChange.php

```diff
@@ -0,0 +1,48 @@
+<?php
+
+use CRM_HRLeaveAndAbsences_BAO_LeaveBalanceChange as LeaveBalanceChange;
+use CRM_HRLeaveAndAbsences_BAO_LeaveRequest as LeaveRequest;
+
+class CRM_HRLeaveAndAbsences_Service_LeaveBalanceChange {
+
+  /**
+   * Creates LeaveBalanceChange instances for each of the dates of the given
+   * LeaveRequest
+   */
+  public function createForLeaveRequest(LeaveRequest $leaveRequest) {
+    $balanceChanges = $this->calculateBalanceChanges($leaveRequest);
+
+    $dates = $leaveRequest->getDates();
+
+    $balanceChangeTypes = array_flip(LeaveBalanceChange::buildOptions('type_id', 'validate'));
+    foreach($dates as $date) {
+      foreach($balanceChanges['breakdown'] as $balanceChange) {
+        if($balanceChange['date'] == $date->date) {
+          LeaveBalanceChange::create([
+            'source_id' => $date->id,
+            'source_type' => LeaveBalanceChange::SOURCE_LEAVE_REQUEST_DAY,
+            'amount' => $balanceChange['amount'],
+            'type_id' => $balanceChangeTypes['debit']
+          ]);
+        }
+      }
+    }
+  }
+
+  /**
+   * Calculates the balance changes for each of the LeaveRequest dates
+   *
+   * @param \CRM_HRLeaveAndAbsences_BAO_LeaveRequest $leaveRequest
+   *
+   * @return array
+   */
+  private function calculateBalanceChanges(LeaveRequest $leaveRequest) {
+    return LeaveRequest::calculateBalanceChange(
+      $leaveRequest->contact_id,
+      new DateTime($leaveRequest->from_date),
+      $leaveRequest->from_date_type,
+      empty($leaveRequest->to_date) ? null : new DateTime($leaveRequest->to_date),
+      $leaveRequest->to_date_type
```

Comment:
No. The check for to_date is required because:

- calculateBalanceChange expects a DateTime object for that param
- We don't want to pass a null to the DateTime constructor

On the other hand, calculateBalanceChange expects a string for the to_date_type param, including null, so there's no need to check for anything